feat(config): support user configured env (#27028)
renovatebot/renovate: docs/usage/configuration-options.md

For more information on how to use secrets for private packages, read [Private package support](./getting-started/private-packages.md).

## env

> This option allows users to specify explicit environment variables and their corresponding values.
> 

Reviewer: ```suggestion
This option allows users to specify explicit environment variables and their corresponding values.

<!-- prettier-ignore -->
!!! note
    This option is only available to users if the bot administrator has configured an [`allowedEnv`](./self-hosted-configuration.md#allowedEnv) list permitting the env names you wish to use.

```

Behavior:

> - Currently this option is only applicable to child processes, specifically within the `updateArtifcats()` function
> - Values defined in the env configuration take precedence over corresponding environment variables, including those from `customEnvVariables` and `process.env`
> - Before adding variables to the env configuration, ensure they are included in the [`allowedEnv`](./self-hosted-configuration.md#allowedEnv) list within the admin configuration
>   Example:

Reviewer: ```suggestion
- This option is only applicable to when executing package manager commands, specifically within the `updateArtifacts()` function
- Values defined in the env configuration take precedence over corresponding environment variables, including those from `customEnvVariables` and `process.env`
```

Reviewer 2: I think we should keep this tip, if it's still relevant?

> - Before adding variables to the env configuration, ensure they are included in the [`allowedEnv`](./self-hosted-configuration.md#allowedEnv) list within the admin configuration

> ```json

Reviewer: ```suggestion
```json title="Example renovate.json with env configuration"
```

For more information on how to use secrets for private packages, read [Private package support](./getting-started/private-packages.md).

## env

This option allows users to specify explicit environment variables and their corresponding values.

<!-- prettier-ignore -->

> !!! note

Reviewer 2: ```suggestion
!!! warning
```

Should we bump this to `warning`? I think a note is easier to ignore than a warning type message. Reading the current docs in this PR, it looks like the bot admin _must_ do some other config first, before end-users can use the `env` config option. So we should shout a bit louder about the required steps, I think.

>     This option is only available to users if the bot administrator has configured an [`allowedEnv`](./self-hosted-configuration.md#allowedEnv) list permitting the env names you wish to use.

Reviewer 2: ```suggestion
    The bot administrator must configure a list of allowed environment names in the  an [`allowedEnv`](./self-hosted-configuration.md#allowedEnv) config option, before users can use those allowed names in the `env` option.
```

How about something like this? I'm trying to explain:

- That bot admin must set allowed names in the `allowedEnv` list
- That users can only use names from the _allowed_ list in their `env` config option

Behavior:

> - This option is only applicable to when executing package manager commands, specifically within the `updateArtifacts()` function

Reviewer 2: ```suggestion
- This option only applies when Renovate executes package manager commands, within the `updateArtifacts()` function
```

How about something like this?

> - This option only applies when Renovate executes package manager commands, within the `updateArtifacts()` function

Reviewer 2: ```suggestion
- This option only applies when Renovate runs package manager commands, within the `updateArtifacts()` function
```

I think we can simplify with `runs` here.

> - Values defined in the env configuration take precedence over corresponding environment variables, including those from `customEnvVariables` and `process.env`

Reviewer 2: ```suggestion
- Values set in the env configuration override corresponding environment variables, including those from `customEnvVariables` and `process.env`
```

`Take precedence over` is difficult English to me. I think a word like `override` is easier.
